cana/boolean_node.py

```diff
@@ -139,6 +156,120 @@ def from_output_list(self, outputs=list(), *args, **kwargs):
+            if required_node_bias > max_achievable_bias:
+                if verbose:
+                    print(f"Required Node Bias is greater than the maximum achievable bias ({max_achievable_bias}) of the node. Generating with the maximum achievable bias.")
+                required_node_bias = max_achievable_bias
+            
+            elif required_node_bias < min_achievable_bias:
+                min_achievable_bias = generated_node.bias(verbose=False)
+                if verbose:
+                    print(f"Required Node Bias is lower than the minimum achievable bias (bias = {min_achievable_bias}) of the node. Generating with the minimum achievable bias.")
+                required_node_bias = min_achievable_bias
+            
+            # Fill the missing output values to achieve the required bias as closely as possible.
+            required_ones = int(required_node_bias * 2**generated_node.k) # recalculating in case the required bias was adjusted in the above steps.
+            ones_to_be_generated = required_ones - current_ones
+            number_of_missing_values = generated_node.outputs.count('?')  
+
+            missing_output_values = ['1'] * ones_to_be_generated + ['0'] * (number_of_missing_values - ones_to_be_generated) # creating a shuffled list of 1 and 0 to replace the '?' with the right ratio required to achieve the required bias.
+            random.shuffle(missing_output_values)
+            generated_node.outputs = [missing_output_values.pop() if output== '?' else output for output in generated_node.outputs]
+
+            if verbose:
+                print(f"Generated the node with a bias of {generated_node.bias(verbose=False)}. This is the closest bias less than or equal to the required bias of {required_node_bias}.")
+
+        elif fill_missing_output_randomly:
+            # Replace '?' in generated_node.outputs with 0 or 1 randomly
+            generated_node.outputs = [random.choice(['0', '1']) if output == '?' else output for output in generated_node.outputs]
+        
+        elif required_effective_connectivity is not None:
+            generated_outputs = generated_node.outputs.copy()
+            missing_output_indices = [i for i, x in enumerate(generated_outputs) if x == '?']
+            # print(f"Missing output indices = {missing_output_indices}." if verbose else None)
+
+            missing_output_count = generated_outputs.count('?')
+            # print(f"No. of '?' in output = {missing_output_count}.")
+            permutations = list(product(*[('0', '1')] * (missing_output_count)))
+            # print(permutations)  
+            generated_node_permutations = [None] * len(permutations)
+
+            for count, permutation in enumerate(permutations):
+                for i, index in enumerate(missing_output_indices):
+                    generated_outputs[index] = permutation[i]
+                generated_node_permutations[count] = BooleanNode.from_output_list(generated_outputs)
```

Comment:
refactor `generated_node_permutations` into its own function. It will be useful all on its own (probably)